addons/sale/migrations/11.0.1.1/post-migration.py

```diff
@@ -19,9 +20,107 @@ def activate_proforma(env):
         employee_group.write({'implied_ids': [(4, proforma_group.id)]})
 
 
+@openupgrade.logging()
+def fill_sale_order_line_computed(env):
+    """Compute through SQL sale.order.line fields for speeding up migration."""
+    # We first set everything to 0 in both fields
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE sale_order_line sol
+        SET amt_invoiced = 0.00,
+            amt_to_invoice = 0.00
+        """,
+    )
+    # Then we set the invoiced amount for the order lines with related invoice
+    # lines and same currency
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE sale_order_line sol
+        SET amt_invoiced = sub.amount
+        FROM (
+            SELECT rel.order_line_id,
+                ail.currency_id,
+                SUM(ail.price_total) AS amount
+            FROM sale_order_line_invoice_rel rel
+                INNER JOIN account_invoice_line ail ON
+                    rel.invoice_line_id = ail.id
+                INNER JOIN account_invoice ai ON ai.id = ail.invoice_id
+            WHERE ai.state IN ('open', 'paid')
+            GROUP BY rel.order_line_id,
+                ail.currency_id
+        ) AS sub
+        WHERE sol.id = sub.order_line_id
+            AND sol.currency_id = sub.currency_id
+        """,
+    )
+    # Now we set the amt_to_invoice with this simplified computation, but
+    # before discounting refunded invoiced. We apply a fixed rounding for
+    # avoiding resolution problems. It also uses a trick using the same table
+    # 2 times for avoiding to perform a subquery.
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE sale_order_line sol
+        SET amt_to_invoice = ROUND(
+            (CASE WHEN COALESCE(pt.invoice_policy, 'x') = 'delivery'
+             THEN sol.price_reduce_taxinc * sol.qty_delivered
+             ELSE sol.price_total END) - sol.amt_invoiced END, 5
+        )
+        FROM sale_order_line sol2
+        LEFT JOIN product_product pp ON pp.id = sol2.product_id
+        LEFT JOIN product_template pt ON pt.id = pp.product_tmpl_id
+        WHERE sol2.id = sol.id
+            AND sol.state IN ('sale', 'done')
+        """,
+    )
+    # We have to do this second query that reduces the invoiced amount due to
+    # the refunded amount as the computation mixes several tables that can't
+    # be combined together. The computation is not exact, as we reduce amount
+    # for all lines with same product, but it's not even correct on upstream,
+    # as there's no real link.
+    # See https://github.com/odoo/odoo/pull/26054#pullrequestreview-143472733
+    openupgrade.logged_query(
```

Comment:
Some time ago, Odoo added this commit https://github.com/odoo/odoo/commit/c05abb2db837138ef784034aa8bab2fe2e014206. Should we do a change in the query?

Reply:
Well, there's a slight difference in the computation algorithm through SQL and through ORM: if there are multiple lines with the same product, with SQL, the amount is discounted on each line while in ORM only once, but it's not possible to fine-tune better and it's a very edge case. What do you think?

Reply:
Ok! :+1: